src/include/pbs_sched.h

```diff
@@ -83,22 +83,21 @@ extern void *sched_attr_idx;
 extern attribute_def sched_attr_def[];
 
 typedef struct pbs_sched {
-	pbs_list_link	sc_link;		/* forward/backward links */
-	int scheduler_sock;
-	int scheduler_sock2;
-	int svr_do_schedule;
-	int svr_do_sched_high;
-	pbs_net_t pbs_scheduler_addr;
-	unsigned int pbs_scheduler_port;
-	time_t sch_next_schedule;		/* when to next run scheduler cycle */
-	char sc_name[PBS_MAXSCHEDNAME + 1];
-	struct preempt_ordering preempt_order[PREEMPT_ORDER_MAX + 1];
-	/* sched object's attributes  */
-	attribute sch_attr[SCHED_ATR_LAST];
-	short newobj;
+	pbs_list_link sc_link;					      /* link to all scheds known to server */
+	int sc_tmp_primary_conn;				      /* this is just to hold primary connection fd till secondary connection comes */
```

Comment:
No, we can't without performance penalty, as we have to go though each connection in server, try to find respective sched (aka this struct) by matching primary connection (aka loop through each sched in server), then if get sched struct then match sched name with name came in secondary connection.

Reply:
@bhroam true number of sched will be less but number of conn in server can be high if many clients are connected to servers, and will have to loop through each connection -> loop though each sched per connection, so it will be huge performance penalty. And we can avoid those by just using one temporary variable so why not?
Also if assign sc_primary_conn then find_sched_from_sock which finds sched by comparing primary_conn or secondary conn will get sched which is not yet connected fully. so it will cause problem in wherever it is used.

Reply:
Ahh, you are looping through the connections.  Yes, that will slow things down.  As for the temp variable, I am not sure I find a problem with storing the primary connection in the primary connection variable.  Finding a scheduler by the primary sock shouldn't be a problem unless you are trying to run a cycle.  You can easily add a check when you are about to run a cycle that both the primary and secondary descriptors are there before you start it.  Otherwise, I think it would be OK to find it.  It IS the scheduler's descriptor.  If you don't like this, just modify find_sched_by_sock() to make sure there is a valid primary and secondary connection before you search.

Reply:
@bhroam initially when I started developing this changes, I used primary connection variable only (earlier it was named as 'scheduler_sock'), and when I run regression I show failures and and while debugging I found that root cause was early assignment of primary connection, so I used another tmp variable, but if you are saying that it should cause problem then I will give a try again without tmp variable and see whether it works or not.

Reply:
I'm really surprised there was a problem.  The only time in the code we should have had a socket to search for was if a cycle was active.  When we're registering connections, we can't have a cycle active, so none of that code should be executing.  If it is, then we're assuming that if we can find a scheduler by a socket, then a cycle is active. If that is the case, that code is now wrong.  In any case, I don't like the idea of the temp variable.  I just don't think it is necessary.  If you truly find that it is, then so be it, but I'd rather find out why things weren't working right before.

Reply:
> The only time in the code we should have had a socket to search for was if a cycle was active. When we're registering connections, we can't have a cycle active, so none of that code should be executing. If it is, then we're assuming that if we can find a scheduler by a socket, then a cycle is active. If that is the case, that code is now wrong.

Yes, this is the case. Since earlier server was connecting to sched so we had sockets only when cycle was active, but now since we will always have socket no matter whether cycle is active or not. But I agree with you that code shouldn't assume then if it finds sched by a socket then it shouldn't assume that cycle is active, I will look into this and try to fix it.